sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -1209,6 +1209,12 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
         case p @ (_: ResolvedHint | _: LeafNode | _: Repartition | _: SubqueryAlias) =>
           p.children.foreach(child => checkPlan(child, aggregated, canContainOuter))
 
+        case p @ (_ : Union) =>
+          // Set operations (e.g. UNION) containing correlated values are only supported
+          // with decorrelateInnerQueryEnabled.
+          val childCanContainOuter = canContainOuter && SQLConf.get.decorrelateInnerQueryEnabled
```

Comment:
I think we also need to check if it's inside a scalar or lateral subquery. I don't think we support this in IN/EXISTS subquery.

Reply:
Added the check and tests for EXISTS. Correlation under IN is supported via RewritePredicateSubquery to rewrite to semi join. I found existing tests for IN at in-set-operations.sql.